python/ql/lib/semmle/python/security/dataflow/UnicodeBypassValidationQuery.qll

```diff
@@ -0,0 +1,73 @@
+/**
+ * Provides a taint-tracking configuration for detecting "Unicode transformation mishandling" vulnerabilities.
+ */
+
+private import python
+import semmle.python.ApiGraphs
+import semmle.python.Concepts
+import semmle.python.dataflow.new.DataFlow
+import semmle.python.dataflow.new.internal.DataFlowPublic
+import semmle.python.dataflow.new.TaintTracking
+import semmle.python.dataflow.new.internal.TaintTrackingPrivate
+import semmle.python.dataflow.new.RemoteFlowSources
+import UnicodeBypassValidationCustomizations::UnicodeBypassValidation
+
+/** A state signifying that a logical validation has not been performed. */
+class PreValidation extends DataFlow::FlowState {
+  PreValidation() { this = "PreValidation" }
+}
+
+/** A state signifying that a logical validation has been performed. */
+class PostValidation extends DataFlow::FlowState {
+  PostValidation() { this = "PostValidation" }
+}
+
+/**
+ * A taint-tracking configuration for detecting "Unicode transformation mishandling" vulnerabilities.
+ *
+ * This configuration uses two flow states, `PreValidation` and `PostValidation`,
+ * to track the requirement that a logical validation has been performed before the Unicode Transformation.
+ */
+class Configuration extends TaintTracking::Configuration {
+  Configuration() { this = "UnicodeBypassValidation" }
+
+  override predicate isSource(DataFlow::Node source, DataFlow::FlowState state) {
+    source instanceof RemoteFlowSource and state instanceof PreValidation
+  }
+
+  override predicate isAdditionalTaintStep(
+    DataFlow::Node nodeFrom, DataFlow::FlowState stateFrom, DataFlow::Node nodeTo,
+    DataFlow::FlowState stateTo
+  ) {
+    (
+      exists(Escaping escaping | nodeFrom = escaping.getAnInput() and nodeTo = escaping.getOutput())
+      or
+      exists(RegexExecution re | nodeFrom = re.getString() and nodeTo = re)
+      or
+      stringManipulation(nodeFrom, nodeTo)
```

Comment:
Indeed, I used `stringManipulation()` because of the second part `// String methods`. This part of interest [TaintTrackingPrivate.qll#L104-L135](https://github.com/github/codeql/blob/58f4b7696da9d02514a3fd09c82e41541d8165d8/python/ql/lib/semmle/python/dataflow/new/internal/TaintTrackingPrivate.qll#L104-L135). 

Do you want me to rip that code and inject it within the UBV Query, please?

Reply:
I believe that is the right thing to be done commit [be3f59a](https://github.com/github/codeql/pull/12991/commits/be3f59afab9725ecc6bbf900989108fd2fdc1a27).

Reply:
No problem; I would suggest `stringManipulation()` to be somehow part of CodeQL Concepts.